src/helper/remote_config/asm_listener.cpp

```diff
@@ -0,0 +1,77 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+#include "asm_listener.hpp"
+#include "../json_helper.hpp"
+#include "exception.hpp"
+#include "spdlog/spdlog.h"
+#include <optional>
+#include <rapidjson/document.h>
+#include <rapidjson/rapidjson.h>
+
+namespace dds::remote_config {
+
+namespace {
+
+// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+void merge_arrays(rapidjson::Value &destination, rapidjson::Value &source,
+    rapidjson::Value::AllocatorType &allocator)
+{
+    for (auto *it = source.Begin(); it != source.End(); ++it) {
+        destination.PushBack(*it, allocator);
+    }
+}
+
+} // namespace
+
+void asm_listener::on_update(const config &config)
+{
+    rapidjson::Document doc(rapidjson::kObjectType, &ruleset_.GetAllocator());
+    if (!json_helper::get_json_base64_encoded_content(config.contents, doc)) {
+        throw error_applying_config("Invalid config contents");
+    }
+
+    for (auto destination_it = ruleset_.MemberBegin();
```

Comment:
In case of something calling to this method without having called first to init, it will result on a segmentation fault i think

Reply:
The new pattern for listeners requires calling `init` before `update` / `unapply` and `commit` after, I understand that a partial state could cause some issues, although it shouldn't really happen aside from in tests because the listeners are always called in the same place.  We can add extra guardrails I suppose, but it's unlikely an incorrect use would ever work considering in debug mode rapidjson asserts.